Index: ../code/trunk/pp3/au-peopleprofile-sap.extent.uilib-web/src/main/uilib/sap/extent/uilib/util/EditFormGenerator.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- ../code/trunk/pp3/au-peopleprofile-sap.extent.uilib-web/src/main/uilib/sap/extent/uilib/util/EditFormGenerator.js	(revision 888338)
+++ ../code/trunk/pp3/au-peopleprofile-sap.extent.uilib-web/src/main/uilib/sap/extent/uilib/util/EditFormGenerator.js	(revision )
@@ -1269,17 +1269,23 @@
         return oDetailsButton;
     } else if (oMeta.id === sap.extent.uilib.util.EditFormGenerator.DETAILS_HEADER_FAKE_ITEM) {
         var oData = oContext.getModel().getData();
+        var indexOfHeader = 0;
+        while(indexOfHeader < (oData.metadataAsList.fields.length-1) &&  oData.metadataAsList.fields[indexOfHeader].type =="DATE"){
+            indexOfHeader ++;
+        }
         if (oData.dataList && oData.dataList.length > 0 && oData.dataList[0].v && oData.dataList[0].v.length > 0){
             var oColumnHeader = new sap.m.Text({
                 text: {
-                    parts: [oContext.getPath().split("v")[0]+"v/0/value", oContext.getPath().split("v")[0]+"v/0/meta"],
+                    parts: [oContext.getPath().split("v")[0]+"v/"+indexOfHeader+"/value", oContext.getPath().split("v")[0]+"v/"+indexOfHeader+"/meta"],
                     formatter: function (oValue, oMeta) {
                         if (oValue) {
-                            return EditFormGeneratorHelper.getLabelValue(oValue, oMeta);
+                           return EditFormGeneratorHelper.getLabelValue(oValue, oMeta);
                         }
                     }
-                }
+                },
+                maxLines:1
             });
+            oColumnHeader.addStyleClass("sfPeopleProfileEditFormHeaderFake");
         return oColumnHeader;
         }
     } else {
@@ -1620,6 +1626,9 @@
 //Column Header on table pop-in
 sap.extent.uilib.util.EditFormGenerator.prototype._addHeaderOnPopIn = function(oModelAsList) {
     if (oModelAsList && oModelAsList.dataList && oModelAsList.metadataAsList && oModelAsList.metadataAsList.fields) {
+        if(oModelAsList.blockType !== "BACKGROUND_BLOCK"){
+            return;
+        }
         for ( var j = 0; j < oModelAsList.dataList.length; j++) {
             // add header for repeating blocks, i.e. when table is popped in
                 var vCountofFake = 0;
Index: ../code/trunk/pp3/au-peopleprofile-sap.sf.peopleprofile-web/src/main/uilib/sap/sf/themes/sap_bluecrystal/app.less
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- ../code/trunk/pp3/au-peopleprofile-sap.sf.peopleprofile-web/src/main/uilib/sap/sf/themes/sap_bluecrystal/app.less	(revision 888338)
+++ ../code/trunk/pp3/au-peopleprofile-sap.sf.peopleprofile-web/src/main/uilib/sap/sf/themes/sap_bluecrystal/app.less	(revision )
@@ -718,6 +718,10 @@
   padding-left: 0rem;
 }
 
+.sfPeopleProfileEditFormHeaderFake{
+  max-width: 60%;
+}
+
 .sfPeopleProfileSkillsTablePhonePadding{
   padding-bottom:0.5rem;
 }
\ No newline at end of file
Index: ../code/trunk/pp3/au-peopleprofile-sap.sf.peopleprofile-web/src/main/webapp/ui/peopleprofile/resources/sap/sf/peopleprofile/model/DWRConverter.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- ../code/trunk/pp3/au-peopleprofile-sap.sf.peopleprofile-web/src/main/webapp/ui/peopleprofile/resources/sap/sf/peopleprofile/model/DWRConverter.js	(revision 888338)
+++ ../code/trunk/pp3/au-peopleprofile-sap.sf.peopleprofile-web/src/main/webapp/ui/peopleprofile/resources/sap/sf/peopleprofile/model/DWRConverter.js	(revision )
@@ -44,7 +44,7 @@
         configParamsInfo: "object",
         changedData: "array"
     };
-    
+
     oDataJSON.effectiveDate = dwrJSON.effectiveAsOf;
 
     sap.sf.peopleprofile.model.DWRUtilsAdapter.strip(dwrJSON, toStrip, oDataJSON);
@@ -115,7 +115,7 @@
                     }
                 };
             }
-            
+
             if (dwrJSON.historyEnabled) {
                 if(oDataJSON.portletID === "COST_DISTRIBUTION_INFO"){
                     oDataJSONFieldGroups.isMDFHistory = true;
@@ -311,7 +311,7 @@
             }
             oDataJSONSection.data.fields[vaueObj.id].value.push(valueDefault);
             oDataJSONSection.data.fields[vaueObj.id].oldValue.push(valueOld);
-            
+
             if (vaueObj.otherValue) {
                 var valueOther = vaueObj.otherValue;
                 if (sap.sf.peopleprofile.model.DWRJSONUtils.isJSON(valueOther)) {
@@ -415,13 +415,13 @@
     for (var fieldIndex = 0, fieldLength = dwrMetaJson.length; fieldIndex < fieldLength; fieldIndex++) {
         var fieldMeta = {};
         var fieldMetaConf = dwrMetaJson[fieldIndex];
-        
+
         if(!fieldMetaConf.hasOwnProperty("writable") && !fieldMetaConf.hasOwnProperty("columnEditable")){
         	fieldMetaConf.writable = true;
         }else{
-        	fieldMetaConf.writable = fieldMetaConf.writable || !!fieldMetaConf.columnEditable;            
+        	fieldMetaConf.writable = fieldMetaConf.writable || !!fieldMetaConf.columnEditable;
         }
-        
+
         if(typeof fieldMetaConf.editable === "boolean"){
             fieldMetaConf.writable = fieldMetaConf.editable;
         }
@@ -465,7 +465,7 @@
 
         fieldMeta.serverSentType = fieldMeta.type;
         fieldMeta.type = fieldMeta.type.toUpperCase();
-        if (fieldMeta.type === 'PICKLISTBEAN' || fieldMeta.type === 'SELECT') {        	
+        if (fieldMeta.type === 'PICKLISTBEAN' || fieldMeta.type === 'SELECT') {
             if (fieldMetaConf.ajaxSelect) {
                 fieldMeta.type = 'PICKLISTAUTOCOMPLETE';
             } else {
@@ -671,7 +671,7 @@
         }
         var orderPos = 1;
         for (var technicalIndex = 0, technicalLength = techInfo.length; technicalIndex < technicalLength; technicalIndex++) {
-            var rowData = {}, emptyRow = null;            
+            var rowData = {}, emptyRow = null;
             var eachTechInfo = techInfo[technicalIndex];
             if (eachTechInfo.actionType === "Delete") {
                 eachItem.deletedRows.push(eachTechInfo.id);
@@ -968,7 +968,9 @@
             "group0":{
                 "fields":[],
                 "renderStyle":"table"
-            }}
+            }},
+        blockType:isTrend?"TREND_BLOCK":"BACKGROUND_BLOCK"
+
     };
     if(isTrend){
         oDataJSONFieldGroups['isTrend']=isTrend
@@ -1350,7 +1352,7 @@
                 valueObj.sections = sap.sf.peopleprofile.model.DWRConverter.convertODataToDWRData(wizardStepPortlet);
             }
             valueObj.portletId = wizardStepPortlet.portletID;
-            sap.sf.peopleprofile.model.DWRUtilsAdapter.strip(wizardStepPortlet, 
+            sap.sf.peopleprofile.model.DWRUtilsAdapter.strip(wizardStepPortlet,
                 sap.sf.peopleprofile.model.DWRConverter.oDataToDWRDataStrip(wizardStepPortlet), valueObj);
             wizardPage.wizardStepPortlets[index] = valueObj;
         }
\ No newline at end of file
